cotainr/pack.py

```diff
@@ -279,6 +280,37 @@ def _display_miniforge_license_for_acceptance(self, *, installer_path):
                 "No license seems to be displayed by the Miniforge installer."
             )
 
+    def _get_install_script(self):
+        """
+        Determine the Miniforge installer to be downloaded based on system information.
```

Comment:
Isn't the Miniforge installer run within the container? The host OS shouldn't matter...

Reply:
You still have to choose a Miniforge installer matching the architecture of your base image. Though it is an interesting thought to have only x86_64 containers and run those on aarch64 based systems, e.g. an [Nvidia GH200](https://www.nvidia.com/en-us/data-center/grace-hopper-superchip/) based system, I don't think it makes much sense from a performance point of view since you would have to use some layer of emulation when running the x86_64 container on aarch64.

More background information in https://github.com/apptainer/singularity/issues/4244#issuecomment-523047798 and https://sylabs.io/2022/07/cross-architecture-containers-with-singularityce-pro/.

Since the HPC ecosystem is increasingly adopting aarch64 CPUs, I think at some point in the near future, we need to natively support aarch64 container builds in cotainr.

Reply:
Sure. But looking at the architecture and OS of the host machine is not the way to go, I don't think - I routinely build amd64 Linux Docker images on my arm64 macOS machine.

Reply:
Ahh, I think, I get your point now. You want the option to build aarch64 containers on a x86_64 machine, or vice versa, right? That's a bit more ambitious than what we have been pursuing so far with cotainr, but not entirely unreasonable. Cotainr originated from a desire to be able to easily build containers in user space on HPC systems with no "special" kernel features, e.g. user namespaces, virtualization, etc. Basically, we just needed something that could build containers matching the architecture of the host HPC system.

Where do you usually run cotainr? On your laptop? On an HPC system? In the cloud? Or elsewhere? And what is your typical use case that cotainr supports?

Reply:
> Basically, we just needed something that could build containers matching the architecture of the host HPC system.

That's all fine and dandy (and Cotainr is much better than certain other options), but what this PR is doing is downloading e.g. a macOS or Windows installer for Miniforge if the host system happens to be that - and as I see it, since the images that are being built are all Linux based, trying to run a macOS or Windows Miniforge installer within that image wouldn't fly. 

> Where do you usually run cotainr? On your laptop? On an HPC system?

I would run it on LUMI if #55 got merged 😉, and I run it on my laptop for development with a containerized amd64 Singularity.

Reply:
> > Basically, we just needed something that could build containers matching the architecture of the host HPC system.
> 
> That's all fine and dandy (and Cotainr is much better than certain other options), but what this PR is doing is downloading e.g. a macOS or Windows installer for Miniforge if the host system happens to be that - and as I see it, since the images that are being built are all Linux based, trying to run a macOS or Windows Miniforge installer within that image wouldn't fly.

I see, you are right that the current implementation doesn't make much sense since it relies on the architecture of the host which definitely isn't what we want. I should have looked more closely at the code to realize this earlier on. Sorry for the confusion. I guess, if the check for `platform.machine()` was done inside the container, it would have ensured a match between base image and Miniforge installer, though I am not entirely sure how `platform` works when invoked within a container.

> 
> > Where do you usually run cotainr? On your laptop? On an HPC system?
> 
> I would run it on LUMI if #55 got merged 😉, and I run it on my laptop for development with a containerized amd64 Singularity.

I am very interested in learning more about this, so I will pick up on it in the discussion in #55.

Reply:
> I guess, if the check for `platform.machine()` was done inside the container, it would have ensured a match between base image and Miniforge installer, though I am not entirely sure how `platform` works when invoked within a container.

It would work fine, but it would require Python to be installed within the container before Miniforge is installed, which could be an issue. Instead, one could

* introspect the output of [`arch`](https://man7.org/linux/man-pages/man1/arch.1.html) or [`uname -m`](https://linux.die.net/man/1/uname) (`aarch64` on Linux-on-ARM64, `x86_64` on Linux-on-AMD64, `arm64` on macOS-on-Apple-Silicon),
* or simply accept the target architecture as a command line parameter for now.

Reply:
Considering we just want to build containers for the same platform i would just make the different versions call the linux script as the container we are pulling is a linux. Then we can extend with an architecture command line parameter when/if we set up pulling containers of different architectures with cotainr.